lib/mongoid/traversable.rb

```diff
@@ -28,6 +28,8 @@ def discriminator_key=(value)
           raise Errors::InvalidDiscriminatorKeyTarget.new(self, self.superclass)
         end
 
+        self._clear_types if self.respond_to?(:_clear_types)
```

Comment:
Why is the respond_to check here?

Reply:
This function is often called before _clear_types (or even _types) is declared resulting in failure... I figured since _types doesn't even exist yet then calling _clear_types doesn't do anything at all.